routing: move call to PruneGraph outside of the collating loop

Crypt-iQ opened the pull request:
This commit moves the call to `PruneGraph` outside of the loop that collates all of the `spentOutputs` in `syncGraphWithChain`. With this change, if a node has been offline for a long period of time, resyncing with the chain no longer takes up as much memory (1MB vs 200MB in some cases) or time. Previously, PruneGraph was called for every block and allocated a very large map further down in the pruneGraphNodes function. Now, pruneGraphNodes is only called once.

I found this by profiling lnd and noticed that it allocated a huge amount of memory in the `pruneGraphNodes` function.  This was after my node was offline for several hundred blocks.  The `syncGraphWithChain` function also took ~20 seconds.
```
ROUTINE ======================== github.com/lightningnetwork/lnd/channeldb.(*ChannelGraph).pruneGraphNodes.func1 in /Users/nsa/go/src/github.com/lightningnetwork/lnd/channeldb/graph.go
  185.88MB   185.88MB (flat, cum) 15.81% of Total
         .          .    934:			return nil
         .          .    935:		}
         .          .    936:
         .          .    937:		var nodePub [33]byte
         .          .    938:		copy(nodePub[:], pubKey)
  185.88MB   185.88MB    939:		nodeRefCounts[nodePub] = 0
         .          .    940:
         .          .    941:		return nil
         .          .    942:	})
         .          .    943:	if err != nil {
         .          .    944:		return err
```

After moving `PruneGraph` out of the loop, initial sync of ~300 blocks took ~1 second and only ~1MB on the heap :)

```
ROUTINE ======================== github.com/lightningnetwork/lnd/channeldb.(*ChannelGraph).pruneGraphNodes in /Users/nsa/go/src/github.com/lightningnetwork/lnd/channeldb/graph.go
         0     1.05MB (flat, cum)  0.13% of Total
         .          .    924:
         .          .    925:	// We'll use this map to keep count the number of references to a node
         .          .    926:	// in the graph. A node should only be removed once it has no more
         .          .    927:	// references in the graph.
         .          .    928:	nodeRefCounts := make(map[[33]byte]int)
         .     1.05MB    929:	err = nodes.ForEach(func(pubKey, nodeBytes []byte) error {
         .          .    930:		// If this is the source key, then we skip this
         .          .    931:		// iteration as the value for this key is a pubKey
         .          .    932:		// rather than raw node information.
         .          .    933:		if bytes.Equal(pubKey, sourceKey) || len(pubKey) != 33 {
         .          .    934:			return nil
```

```
2019-06-27 13:24:50.588 [INF] CRTR: Prune tip for Channel Graph: height=1565711, hash=000000000000000f871f41667bd527f58e7be40f1d3cc6fd91e98e3fb555130e
2019-06-27 13:24:50.588 [INF] CRTR: Syncing channel graph from height=1565711 (hash=000000000000000f871f41667bd527f58e7be40f1d3cc6fd91e98e3fb555130e) to height=1566009 (hash=00000000000004f825026610b2cc7757c0eaf1d5fd49f0e94cb9c092bec4edc5)
2019-06-27 13:24:51.648 [INF] CHDB: Pruned unconnected node 023d6f31cb13a37c4ede2cd6e0aa119d933ca18fa15244778a72e8ca0232f53539 from channel graph
2019-06-27 13:24:51.648 [INF] CHDB: Pruned unconnected node 0291f510a28f04c76cf18d2c48096f7d5332157443de46e94c188a5e30d51014d4 from channel graph
2019-06-27 13:24:51.648 [INF] CHDB: Pruned unconnected node 030f933df635a4b7ce68cd36f8f25521babe015d1f611775de997d5b35618030ae from channel graph
2019-06-27 13:24:51.649 [INF] CHDB: Pruned 3 unconnected nodes from the channel graph
2019-06-27 13:24:51.683 [INF] CRTR: Graph pruning complete: 4 channels were closed since height 1565711
```

wpaulino commented on routing/router.go:
```diff
@@ -705,33 +706,26 @@ func (r *ChannelRouter) syncGraphWithChain() error {
+	// channel graph, also passing in the best hash+height
+	// so the prune tip can be updated.
+	closedChans, err := r.cfg.Graph.PruneGraph(spentOutputs,
+		bestHash,
```
Style nit: in these cases, we prefer to write this as:

```go
r.cfg.Graph.PruneGraph(
    spentOutputs, bestHash, uint32(bestHeight),
)
```

Crypt-iQ replied:
Fixed

Crypt-iQ commented:
I think here, since the changes are rolled back since the last prune tip via `DisconnectBlockAtHeight`, the chain filter should also be updated to this previous tip.  Otherwise, if there's a re-org, we would only prune the graph for any new blocks and forget about old changes since the last tip.

https://github.com/lightningnetwork/lnd/blob/f4feb66a1fe1a933e798352458a32db1844e5477/routing/router.go#L934-L953

wpaulino commented on routing/router.go:
```diff
@@ -677,10 +677,11 @@ func (r *ChannelRouter) syncGraphWithChain() error {
 	log.Infof("Syncing channel graph from height=%v (hash=%v) to height=%v "+
 		"(hash=%v)", pruneHeight, pruneHash, bestHeight, bestHash)
 
+	var spentOutputs []*wire.OutPoint
```
I believe this was moved in order to gather all of the spent outputs within the height range we're filtering, which are then used within the single `PruneGraph` call.

cfromknecht replied:
yes, i'm referring to the outer for loop

Crypt-iQ commented:
Comments addressed @cfromknecht

cfromknecht commented:
@Crypt-iQ changes lgtm, ready for squash!

Crypt-iQ commented:
@cfromknecht squashed
